libavformat/movenccenc.c

```diff
@@ -83,21 +88,51 @@ static int auxiliary_info_add_subsample(MOVMuxCencContext* ctx,
     return 0;
 }
 
+void ff_mov_cenc_auxiliary_info_reset(MOVMuxCencContext* ctx) {
+    ctx->auxiliary_info_size = 0;
+    ctx->auxiliary_info_entries = 0;
+}
+
 /**
  * Encrypt the input buffer and write using avio_write
+ *
+ * For AES-CBC, the block chain starting with MOVMuxCencContext->aes_cbc_iv is
+ * wholly contained in this function; i.e. do not expect the chain span multiple
+ * calls.
  */
 static void mov_cenc_write_encrypted(MOVMuxCencContext* ctx, AVIOContext *pb,
                                      const uint8_t *buf_in, int size)
 {
-    uint8_t chunk[4096];
+    uint8_t chunk[4000];
     const uint8_t* cur_pos = buf_in;
     int size_left = size;
-    int cur_size;
+    int cur_size, block_count, enc_size, remaining_size;
+    uint8_t aes_cbc_iv[AES_BLOCK_SIZE];
+
+    /* pattern encryption repeats every 10 blocks */
+    av_assert2(sizeof(chunk) % (AES_BLOCK_SIZE * 10) == 0);
 
+    if (ctx->encryption_scheme == MOV_ENC_CENC_AES_CBC_PATTERN) {
+        /* use constant IV and reset the CBC cipher for each sample/subsample */
+        av_aes_reset(ctx->aes_cbc);
+        memcpy(aes_cbc_iv, ctx->aes_cbc_iv, sizeof(aes_cbc_iv));
+    }
+    
     while (size_left > 0) {
         cur_size = FFMIN(size_left, sizeof(chunk));
-        av_aes_ctr_crypt(ctx->aes_ctr, chunk, cur_pos, cur_size);
-        avio_write(pb, chunk, cur_size);
+        if (ctx->encryption_scheme == MOV_ENC_CENC_AES_CTR) {
```

Comment:
Better to use a switch for encryption_scheme and handle valid values, default should error.

Reply:
I'm going to pass on using switches because it complicates the code. The scheme is already validated in movenc.c. And it's unlikely that anyone implements the remaining 2 schemes in the future.